core/types/rlp_backwards_compat.libevm_test.go

```diff
@@ -197,35 +197,27 @@ type cChainBodyExtras struct {
 
 var _ BodyHooks = (*cChainBodyExtras)(nil)
 
-func (e *cChainBodyExtras) AppendRLPFields(b rlp.EncoderBuffer, _ bool) error {
-	b.WriteUint64(uint64(e.Version))
-
-	var data []byte
-	if e.ExtData != nil {
-		data = *e.ExtData
-	}
-	b.WriteBytes(data)
-
-	return nil
+func (e *cChainBodyExtras) RLPFieldsForEncoding(b *Body) ([]any, []any) {
+	// The Avalanche C-Chain uses all of the geth required fields (but none of
+	// the optional ones) so there's no need to explicitly list them. This
+	// pattern might not be ideal for readability but is used here for
+	// demonstrative purposes.
+	//
+	// All new fields will always be tagged as optional for backwards
+	// compatibility so this is safe to do, but only for the required fields.
+	req, _ /*drop all optional*/ := NOOPBodyHooks{}.RLPFieldsForEncoding(b)
+	return append(req, e.Version, e.ExtData), nil
 }
 
-func (e *cChainBodyExtras) DecodeExtraRLPFields(s *rlp.Stream) error {
-	if err := s.Decode(&e.Version); err != nil {
-		return err
-	}
-
-	buf, err := s.Bytes()
-	if err != nil {
-		return err
-	}
-	if len(buf) > 0 {
-		e.ExtData = &buf
-	} else {
-		// Respect the `rlp:"nil"` field tag.
-		e.ExtData = nil
-	}
-
-	return nil
+func (e *cChainBodyExtras) RLPFieldPointersForDecoding(b *Body) ([]any, []any) {
+	// An alternative to the pattern used above is to explicitly list all
+	// fields for better introspection.
+	return []any{
+		&b.Transactions,
+		&b.Uncles,
+		&e.Version,
+		rlp.Nillable(&e.ExtData), // equivalent to `rlp:"nil"`
```

Comment:
if we are adding something like `rlp.Nillable`, does it make sense to also add `rlp.Optional`?
It seems like a better callsite experience than passing back 2 `[]any`

Reply:
I thought of this but then got stuck on what the rest of the API looks like (e.g. what does `EncodeStructField()` accept? Do non-optional fields also have to be wrapped in something?). I'll ponder it a bit more, experiment with the rest of the code, and get back to you.

Reply:
Done. WDYT? It feels a little bit clunky _inside_ the implementation (hence the explanatory comment), but so much clearer here so I completely agree that:

> It seems like a better callsite experience than returning 2 []any

Reply:
Now that we have gone this far, what do you think about adding rlp.Required as well? So the callsite can look like:

```
return rlp.Required(b.Transactions, b.Uncles), rlp.Optional(b.Withdrawals)
```
instead of:
```
return []any{b.Transactions, b.Uncles}, rlp.Optional(b.Withdrawals)
```